### PR TITLE
Validation

### LINKED ISSUE #19: Staticmap: size limit vs tile stitching for high resolution maps
There's an implicit 1280x1280 pixel limit on the size.

For example, creating wallpapers for my Thinkpad X1 Carbon fails with an API error:

    mapbox staticmap --zoom 12 --size 2560 1440 --lat 52.500385 --lon 13.419779 mapbox.emerald out.png

It would be great, to either document that limit and catch the error already at the argparser level or provide transparent stitching, like @zugaldia does in his [wallmapper](https://github.com/zugaldia/wallmapper) (this may be out of scope, though).

### LINKED ISSUE #20: Validation strategy
A general design question: What is our strategy for dealing with validation of input parameters?

The way it works now is
* some known bad parameters are caught by the SDK and raise a python exception (the stacktrace is a bit intimidating and exposes the python underbelly to the CLI world)
* other bad parameters are passed through to the API and the error message is conveyed (actually much cleaner errors, for example the static map size error in #19  returns `Error: {"message":"Height must be between 1-1280."}`) This puts a burden on the API when we could easily catch it before we hit the request.

I'm thinking the best approach is:
1. do as much validation as possible in the SDK and raise a consistent exception type (ValueError seems too general)
2. catch that exception at the CLI and re-raise using either a [`click.UsageError`](http://click.pocoo.org/5/api/#click.UsageError) or a [`click.BadParameter`](http://click.pocoo.org/5/api/#click.BadParameter)

Alternatively, we could do all the validation at the CLI end and let the Python SDK play fast and loose, relying on the API response.

Doing the validation in both spots seems like a bad plan.

/cc @sgillies

### PR DESCRIPTION
**Pull request by perrygeo**
Once https://github.com/mapbox/mapbox-sdk-py/pull/96 and https://github.com/mapbox/mapbox-sdk-py/pull/97 is in master 

Before merging, one last thing:
* [x] unpin the requirement branch, back to master

closes #20 and closes #19

### CODE REVIEW

File: mapboxcli/scripts/directions.py [hunk 1]
```diff
@@ -7,15 +7,17 @@
 
 @click.command(short_help="Routing between waypoints.")
 @click.argument('waypoints', nargs=-1)
-@click.option('--profile', default="mapbox.driving", type=click.Choice([
-              'mapbox.driving', 'mapbox.walking', 'mapbox.cycling']),
-              help="Mapbox direction profile id")
+@click.option('--profile', default="mapbox.driving",
+              help="Mapbox direction profile id, must be one of {0}".format(
+                  mapbox.Directions().valid_profiles))
```
Comment (sgillies) [inline mapboxcli/scripts/directions.py]: How about making this a `click.Choice` populated from `mapbox.Directions().valid_profiles`?

File: mapboxcli/scripts/directions.py [hunk 2]
```diff
@@ -7,15 +7,17 @@
...
-@click.option('--instructions', default="text", type=click.Choice(["text", "html"]),
-              help="Format for route instructions")
-@click.option('--geometry', default="geojson", type=click.Choice([
-              'geojson', 'polyline', 'false']), help="Geometry encoding")
+@click.option('--instructions', default="text",
+              help="Format for route instructions, must be one of {0}".format(
+                  mapbox.Directions().valid_instruction_formats))
+@click.option('--geometry', default="geojson",
+              help="Geometry encoding, must be one of {0}".format(
+                  mapbox.Directions().valid_geom_encoding))
```
Comment (sgillies) [inline mapboxcli/scripts/directions.py]: Same as above. These seem natural for choice, unless I'm missing something (it's late and it's possible).

File: mapboxcli/scripts/directions.py [hunk 3]
```diff
@@ -39,12 +41,17 @@ def directions(ctx, waypoints, geojson, profile, alternatives,
...
+    try:
+        res = service.directions(
+            point_features,
+            steps=steps,
+            alternatives=alternatives,
+            instructions=instructions,
+            geometry=geometry,
+            profile=profile)
+    except mapbox.errors.ValidationError as exc:
+        raise click.BadParameter(str(exc))
```
Comment (sgillies) [inline mapboxcli/scripts/directions.py]: Great. This is exactly how you described it.

File: tests/test_uploads.py [hunk 4]
```diff
@@ -85,3 +85,72 @@ def test_cli_upload(monkeypatch):
...
+         "url": "https://tilestream-tilesets-production.s3.amazonaws.com/_pending/{username}/key.test",
+         "secretAccessKey": "sak.test",
+         "sessionToken": "st.test"}}""".format(username=username)
+
+    responses.add(
+        responses.GET,
+        'https://api.mapbox.com/uploads/v1/{0}/credentials?access_token={1}'.format(username, access_token),
+        match_querystring=True,
+        body=query_body, status=200,
+        content_type='application/json')
```
Comment (sgillies) [inline tests/test_uploads.py]: Maybe not the right place to mention this, but I've been thinking a little about whether and how to make these responses into test fixtures. I guess as soon as we begin to repeat fake responses, we should turn them into a fixture?

**Comment by sgillies** [comment]
@perrygeo looks good to me. Go ahead and merge after making the requirements adjustment.

File: mapboxcli/scripts/directions.py (hunk 1 above)
Comment (perrygeo) [inline mapboxcli/scripts/directions.py]: The only reason I switched to help text was a) following the lead of geocoding and b) I don't like the way the enumerated list looks in the help text when using `Choice`: if you look at the current readme, see how they overlap into the right column? Maybe that's a silly reason - I happy to change it back to `click.Choice` if it makes more sense to you

Comment (sgillies) [inline mapboxcli/scripts/directions.py]: Let's go back to click.Choice then. It looks fine to me.

**Comment by perrygeo** [comment]
The only problem with going to Choices is that I was relying on them to unit test catching the SDK's `ValidationError`. Address it later?

**Comment by sgillies** [comment]
:+1:

**Pull request merged**